core/domain/user_services.py

```diff
@@ -680,36 +680,46 @@ def _save_user_settings(user_settings):
-        gae_id=user_settings.gae_id,
-        email=user_settings.email,
-        role=user_settings.role,
-        username=user_settings.username,
-        normalized_username=user_settings.normalized_username,
-        last_agreed_to_terms=user_settings.last_agreed_to_terms,
-        last_started_state_editor_tutorial=(
+
+    user_settings_dict = {
+        'gae_id': user_settings.gae_id,
+        'email': user_settings.email,
+        'role': user_settings.role,
+        'username': user_settings.username,
+        'normalized_username': user_settings.normalized_username,
+        'last_agreed_to_terms': user_settings.last_agreed_to_terms,
+        'last_started_state_editor_tutorial': (
             user_settings.last_started_state_editor_tutorial),
-        last_started_state_translation_tutorial=(
+        'last_started_state_translation_tutorial': (
             user_settings.last_started_state_translation_tutorial),
-        last_logged_in=user_settings.last_logged_in,
-        last_edited_an_exploration=user_settings.last_edited_an_exploration,
-        last_created_an_exploration=(
+        'last_logged_in': user_settings.last_logged_in,
+        'last_edited_an_exploration': user_settings.last_edited_an_exploration,
+        'last_created_an_exploration': (
             user_settings.last_created_an_exploration),
-        profile_picture_data_url=user_settings.profile_picture_data_url,
-        default_dashboard=user_settings.default_dashboard,
-        creator_dashboard_display_pref=(
+        'profile_picture_data_url': user_settings.profile_picture_data_url,
+        'default_dashboard': user_settings.default_dashboard,
+        'creator_dashboard_display_pref': (
             user_settings.creator_dashboard_display_pref),
-        user_bio=user_settings.user_bio,
-        subject_interests=user_settings.subject_interests,
-        first_contribution_msec=user_settings.first_contribution_msec,
-        preferred_language_codes=user_settings.preferred_language_codes,
-        preferred_site_language_code=(
+        'user_bio': user_settings.user_bio,
+        'subject_interests': user_settings.subject_interests,
+        'first_contribution_msec': user_settings.first_contribution_msec,
+        'preferred_language_codes': user_settings.preferred_language_codes,
+        'preferred_site_language_code': (
             user_settings.preferred_site_language_code),
-        preferred_audio_language_code=(
+        'preferred_audio_language_code': (
             user_settings.preferred_audio_language_code),
-        deleted=user_settings.deleted
-    ).put()
+        'deleted': user_settings.deleted
+    }
+
+    # If user with the given user_id already exists, update that model
+    # with the given user settings, otherwise, create a new one.
+    user_model = user_models.UserSettingsModel.get_by_id(user_settings.user_id)
+    if user_model is not None:
+        user_model.populate(**user_settings_dict)
```

Comment:
In general, I don't believe we use populate() in the domain layer. How are models saved, generally, in other cases, and can we mimic that here instead?

Also, I believe we typically use put() for other models in the codebase. If that is the case, how come they are not affected by this issue?

Reply:
In other cases, a similar approach is taken. For instance, in the [save function](https://github.com/oppia/oppia/blob/38cae409df135c2902246e9ae63049502d201a46/core/domain/collection_services.py#L677) for collections, the existing model is first fetched via the get request, and then it's attributes, like category, title, etc., are changed and the updated model is then put via the commit function.
The problem with the current version of the _save_user_settings function is that it creates a new instance of the UserSettingsModel and then adds the updates to that before putting it (which is why `created_on` was being updated every time some change is made).

Also, the reason for using populate here is so that there is less repetition of code (as suggested by @vojtechjelinek).

Reply:
But then why don't we do something similar here -- fetch the model, make changes to its attributes, and put it? It seems to me that the issue was that we were creating a new model from scratch here and then overwriting.

I also worry about having two different ways to store data (we're introducing a second one here with populate()). It creates inconsistency throughout the codebase. @vojtechjelinek, what do you think?

Reply:
I have no strong opinion about this. The problem with changing the fields directly is that we don't know which fields changed. I think that having a base object for the domain objects that would have some `save` method and would always work the same for all the domain objects might help here.

Reply:
We have discussed this with @iamprayush, let's do it this way for now. In the future we should prepare a design for base domain object that would serve as a common object fro which will the other domain objects inherit and it will handle the common fields (`created_on`, `last_updated_on`, `deleted`).